# libdom: core
# 11 files, 1398 changed lines
# MacSurf fix ids: fixes1005 fixes1042 fixes446 fixes446b fixes446f fixes446g fixes489 fixes490 fixes569
#
# Regenerate: tools/regen-patches.sh && tools/split-patches.py
#
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/libdom/src/core/characterdata_i.h browser/libdom/src/core/characterdata_i.h
--- upstream/libdom/src/core/characterdata_i.h	1970-01-01 01:00:00.000000000 +0100
+++ browser/libdom/src/core/characterdata_i.h	2026-09-20 23:19:17.305551599 +0100
@@ -0,0 +1,127 @@
+/*
+ * This file is part of libdom.
+ * Licensed under the MIT License,
+ *                http://www.opensource.org/licenses/mit-license.php
+ * Copyright 2007 John-Mark Bell <jmb@netsurf-browser.org>
+ */
+
+#ifndef dom_internal_core_characterdata_h_
+#define dom_internal_core_characterdata_h_
+
+#include <dom/core/characterdata.h>
+
+#include "core/node.h"
+
+/**
+ * DOM character data node
+ */
+struct dom_characterdata {
+	struct dom_node_internal base;		/**< Base node */
+};
+
+/* The CharacterData is a intermediate node type, so the following function
+ * may never be used */
+dom_characterdata *_dom_characterdata_create(void);
+dom_exception _dom_characterdata_initialise(struct dom_characterdata *cdata,
+		struct dom_document *doc, dom_node_type type,
+		dom_string *name, dom_string *value);
+
+void _dom_characterdata_finalise(struct dom_characterdata *cdata);
+
+/* The virtual functions for dom_characterdata */
+dom_exception _dom_characterdata_get_data(struct dom_characterdata *cdata,
+		dom_string **data);
+dom_exception _dom_characterdata_set_data(struct dom_characterdata *cdata,
+		dom_string *data);
+dom_exception _dom_characterdata_get_length(struct dom_characterdata *cdata,
+		uint32_t *length);
+dom_exception _dom_characterdata_substring_data(
+		struct dom_characterdata *cdata, uint32_t offset,
+		uint32_t count, dom_string **data);
+dom_exception _dom_characterdata_append_data(struct dom_characterdata *cdata,
+		dom_string *data);
+dom_exception _dom_characterdata_insert_data(struct dom_characterdata *cdata,
+		uint32_t offset, dom_string *data);
+dom_exception _dom_characterdata_delete_data(struct dom_characterdata *cdata,
+		uint32_t offset, uint32_t count);
+dom_exception _dom_characterdata_replace_data(struct dom_characterdata *cdata,
+		uint32_t offset, uint32_t count,
+		dom_string *data);
+dom_exception _dom_characterdata_get_text_content(
+		dom_node_internal *node,
+		dom_string **result);
+dom_exception _dom_characterdata_set_text_content(
+		dom_node_internal *node,
+		dom_string *content);
+
+#define DOM_CHARACTERDATA_VTABLE \
+	_dom_characterdata_get_data, \
+	_dom_characterdata_set_data, \
+	_dom_characterdata_get_length, \
+	_dom_characterdata_substring_data, \
+	_dom_characterdata_append_data, \
+	_dom_characterdata_insert_data, \
+	_dom_characterdata_delete_data, \
+	_dom_characterdata_replace_data 
+
+#define DOM_NODE_VTABLE_CHARACTERDATA \
+	_dom_node_try_destroy, \
+	_dom_node_get_node_name, \
+	_dom_node_get_node_value, \
+	_dom_node_set_node_value, \
+	_dom_node_get_node_type, \
+	_dom_node_get_parent_node, \
+	_dom_node_get_child_nodes, \
+	_dom_node_get_first_child, \
+	_dom_node_get_last_child, \
+	_dom_node_get_previous_sibling, \
+	_dom_node_get_next_sibling, \
+	_dom_node_get_attributes, \
+	_dom_node_get_owner_document, \
+	_dom_node_insert_before, \
+	_dom_node_replace_child, \
+	_dom_node_remove_child, \
+	_dom_node_append_child, \
+	_dom_node_has_child_nodes, \
+	_dom_node_clone_node, \
+	_dom_node_normalize, \
+	_dom_node_is_supported, \
+	_dom_node_get_namespace, \
+	_dom_node_get_prefix, \
+	_dom_node_set_prefix, \
+	_dom_node_get_local_name, \
+	_dom_node_has_attributes, \
+	_dom_node_get_base, \
+	_dom_node_compare_document_position, \
+	_dom_characterdata_get_text_content, /* override */ \
+	_dom_characterdata_set_text_content, /* override */ \
+	_dom_node_is_same, \
+	_dom_node_lookup_prefix, \
+	_dom_node_is_default_namespace, \
+	_dom_node_lookup_namespace, \
+	_dom_node_is_equal, \
+	_dom_node_get_feature, \
+	_dom_node_set_user_data, \
+	_dom_node_get_user_data
+
+/* Following comes the protected vtable 
+ *
+ * Only the _copy function can be used by sub-class of this.
+ */
+void _dom_characterdata_destroy(dom_node_internal *node);
+dom_exception _dom_characterdata_copy(dom_node_internal *old, 
+		dom_node_internal **copy);
+
+#define DOM_CHARACTERDATA_PROTECT_VTABLE \
+	_dom_characterdata_destroy, \
+	_dom_characterdata_copy
+
+extern const struct dom_characterdata_vtable characterdata_vtable;
+
+dom_exception _dom_characterdata_copy_internal(dom_characterdata *old, 
+		dom_characterdata *new);
+#define dom_characterdata_copy_internal(o, n) \
+		_dom_characterdata_copy_internal( \
+		(dom_characterdata *) (o), (dom_characterdata *) (n))
+
+#endif
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/libdom/src/core/document_i.h browser/libdom/src/core/document_i.h
--- upstream/libdom/src/core/document_i.h	1970-01-01 01:00:00.000000000 +0100
+++ browser/libdom/src/core/document_i.h	2026-09-20 23:19:17.306457331 +0100
@@ -0,0 +1,282 @@
+/*
+ * This file is part of libdom.
+ * Licensed under the MIT License,
+ *                http://www.opensource.org/licenses/mit-license.php
+ * Copyright 2007 John-Mark Bell <jmb@netsurf-browser.org>
+ */
+
+#ifndef dom_internal_core_document_h_
+#define dom_internal_core_document_h_
+
+#include <dom/inttypes.h>
+#include <stddef.h>
+
+#include <dom/core/node.h>
+#include <dom/core/attr.h>
+#include <dom/core/cdatasection.h>
+#include <dom/core/comment.h>
+#include <dom/core/document.h>
+#include <dom/core/document_type.h>
+#include <dom/core/doc_fragment.h>
+#include <dom/core/element.h>
+#include <dom/core/entity_ref.h>
+#include <dom/core/pi.h>
+#include <dom/core/text.h>
+#include <dom/core/implementation.h>
+
+#include "core/string.h"
+#include "core/node.h"
+#include "core/nodelist.h"
+
+#include "utils/hashtable.h"
+#include "utils/list.h"
+
+#include "events/document_event.h"
+
+struct dom_doc_nl;
+
+/**
+ * DOM document
+ * This should be protected, because later the HTMLDocument will inherit from
+ * this. 
+ */
+struct dom_document {
+	dom_node_internal base;		/**< Base node */
+
+	struct dom_doc_nl *nodelists;	/**< List of active nodelists */
+
+	dom_string *uri;		/**< The uri of this document */
+
+	struct list_entry pending_nodes;
+			/**< The deletion pending list */
+
+	dom_string *id_name;		/**< The ID attribute's name */
+
+	dom_string *class_string;	/**< The string "class". */
+
+	dom_string *script_string;	/**< The string "script". */
+
+	dom_document_event_internal dei;
+			/**< The DocumentEvent interface */
+	dom_document_quirks_mode quirks;
+				/**< Document is in quirks mode */
+	dom_string *_memo_empty;	/**< The string ''. */
+
+	/* Memoised event strings */
+	dom_string *_memo_domnodeinserted; /**< DOMNodeInserted */
+	dom_string *_memo_domnoderemoved;  /**< DOMNodeRemoved */
+	dom_string *_memo_domnodeinsertedintodocument; /**< DOMNodeInsertedIntoDocument */
+	dom_string *_memo_domnoderemovedfromdocument;  /**< DOMNodeRemovedFromDocument */
+	dom_string *_memo_domattrmodified; /**< DOMAttrModified */
+	dom_string *_memo_domcharacterdatamodified; /**< DOMCharacterDataModified */
+	dom_string *_memo_domsubtreemodified; /**< DOMSubtreeModified */
+
+	uint32_t dispatching_mutation; /**< Mutation event semaphore */
+};
+
+/* Create a DOM document */
+dom_exception _dom_document_create(dom_events_default_action_fetcher daf,
+				   void *daf_ctx,
+				   dom_document **doc);
+
+/* Initialise the document */
+dom_exception _dom_document_initialise(dom_document *doc, 
+		dom_events_default_action_fetcher daf,
+				   void *daf_ctx);
+
+/* Finalise the document */
+bool _dom_document_finalise(dom_document *doc);
+
+/* Begin the virtual functions */
+dom_exception _dom_document_get_doctype(dom_document *doc,
+		dom_document_type **result);
+dom_exception _dom_document_get_implementation(dom_document *doc,
+		dom_implementation **result);
+dom_exception _dom_document_get_document_element(dom_document *doc,
+		dom_element **result);
+dom_exception _dom_document_create_element(dom_document *doc,
+		dom_string *tag_name, dom_element **result);
+dom_exception _dom_document_create_document_fragment(dom_document *doc,
+		dom_document_fragment **result);
+dom_exception _dom_document_create_text_node(dom_document *doc,
+		dom_string *data, dom_text **result);
+dom_exception _dom_document_create_comment(dom_document *doc,
+		dom_string *data, dom_comment **result);
+dom_exception _dom_document_create_cdata_section(dom_document *doc,
+		dom_string *data, dom_cdata_section **result);
+dom_exception _dom_document_create_processing_instruction(
+		dom_document *doc, dom_string *target,
+		dom_string *data,
+		dom_processing_instruction **result);
+dom_exception _dom_document_create_attribute(dom_document *doc,
+		dom_string *name, dom_attr **result);
+dom_exception _dom_document_create_entity_reference(dom_document *doc,
+		dom_string *name,
+		dom_entity_reference **result);
+dom_exception _dom_document_get_elements_by_tag_name(dom_document *doc,
+		dom_string *tagname, dom_nodelist **result);
+dom_exception _dom_document_import_node(dom_document *doc,
+		dom_node *node, bool deep, dom_node **result);
+dom_exception _dom_document_create_element_ns(dom_document *doc,
+		dom_string *namespace, dom_string *qname,
+		dom_element **result);
+dom_exception _dom_document_create_attribute_ns(dom_document *doc,
+		dom_string *namespace, dom_string *qname,
+		dom_attr **result);
+dom_exception _dom_document_get_elements_by_tag_name_ns(
+		dom_document *doc, dom_string *namespace,
+		dom_string *localname, dom_nodelist **result);
+dom_exception _dom_document_get_element_by_id(dom_document *doc,
+		dom_string *id, dom_element **result);
+dom_exception _dom_document_get_input_encoding(dom_document *doc,
+		dom_string **result);
+dom_exception _dom_document_get_xml_encoding(dom_document *doc,
+		dom_string **result);
+dom_exception _dom_document_get_xml_standalone(dom_document *doc,
+		bool *result);
+dom_exception _dom_document_set_xml_standalone(dom_document *doc,
+		bool standalone);
+dom_exception _dom_document_get_xml_version(dom_document *doc,
+		dom_string **result);
+dom_exception _dom_document_set_xml_version(dom_document *doc,
+		dom_string *version);
+dom_exception _dom_document_get_strict_error_checking(
+		dom_document *doc, bool *result);
+dom_exception _dom_document_set_strict_error_checking(
+		dom_document *doc, bool strict);
+dom_exception _dom_document_get_uri(dom_document *doc,
+		dom_string **result);
+dom_exception _dom_document_set_uri(dom_document *doc,
+		dom_string *uri);
+dom_exception _dom_document_adopt_node(dom_document *doc,
+		dom_node *node, dom_node **result);
+dom_exception _dom_document_get_dom_config(dom_document *doc,
+		struct dom_configuration **result);
+dom_exception _dom_document_normalize(dom_document *doc);
+dom_exception _dom_document_rename_node(dom_document *doc,
+		dom_node *node,
+		dom_string *namespace, dom_string *qname,
+		dom_node **result);
+dom_exception _dom_document_get_quirks_mode(dom_document *doc,
+		dom_document_quirks_mode *result);
+dom_exception _dom_document_set_quirks_mode(dom_document *doc,
+		dom_document_quirks_mode result);
+
+
+dom_exception _dom_document_get_text_content(dom_node_internal *node,
+					     dom_string **result);
+dom_exception _dom_document_set_text_content(dom_node_internal *node,
+					     dom_string *content);
+
+#define DOM_DOCUMENT_VTABLE \
+	_dom_document_get_doctype, \
+	_dom_document_get_implementation, \
+	_dom_document_get_document_element, \
+	_dom_document_create_element, \
+	_dom_document_create_document_fragment, \
+	_dom_document_create_text_node, \
+	_dom_document_create_comment, \
+	_dom_document_create_cdata_section, \
+	_dom_document_create_processing_instruction, \
+	_dom_document_create_attribute, \
+	_dom_document_create_entity_reference, \
+	_dom_document_get_elements_by_tag_name, \
+	_dom_document_import_node, \
+	_dom_document_create_element_ns, \
+	_dom_document_create_attribute_ns, \
+	_dom_document_get_elements_by_tag_name_ns, \
+	_dom_document_get_element_by_id, \
+	_dom_document_get_input_encoding, \
+	_dom_document_get_xml_encoding, \
+	_dom_document_get_xml_standalone, \
+	_dom_document_set_xml_standalone, \
+	_dom_document_get_xml_version, \
+	_dom_document_set_xml_version, \
+	_dom_document_get_strict_error_checking, \
+	_dom_document_set_strict_error_checking, \
+	_dom_document_get_uri, \
+	_dom_document_set_uri, \
+	_dom_document_adopt_node, \
+	_dom_document_get_dom_config, \
+	_dom_document_normalize, \
+	_dom_document_rename_node, \
+	_dom_document_get_quirks_mode, \
+	_dom_document_set_quirks_mode
+
+/* End of vtable */
+
+#define DOM_NODE_VTABLE_DOCUMENT \
+	_dom_node_try_destroy, \
+	_dom_node_get_node_name, \
+	_dom_node_get_node_value, \
+	_dom_node_set_node_value, \
+	_dom_node_get_node_type, \
+	_dom_node_get_parent_node, \
+	_dom_node_get_child_nodes, \
+	_dom_node_get_first_child, \
+	_dom_node_get_last_child, \
+	_dom_node_get_previous_sibling, \
+	_dom_node_get_next_sibling, \
+	_dom_node_get_attributes, \
+	_dom_node_get_owner_document, \
+	_dom_node_insert_before, \
+	_dom_node_replace_child, \
+	_dom_node_remove_child, \
+	_dom_node_append_child, \
+	_dom_node_has_child_nodes, \
+	_dom_node_clone_node, \
+	_dom_node_normalize, \
+	_dom_node_is_supported, \
+	_dom_node_get_namespace, \
+	_dom_node_get_prefix, \
+	_dom_node_set_prefix, \
+	_dom_node_get_local_name, \
+	_dom_node_has_attributes, \
+	_dom_node_get_base, \
+	_dom_node_compare_document_position, \
+	_dom_document_get_text_content, \
+	_dom_document_set_text_content, \
+	_dom_node_is_same, \
+	_dom_node_lookup_prefix, \
+	_dom_node_is_default_namespace, \
+	_dom_node_lookup_namespace, \
+	_dom_node_is_equal, \
+	_dom_node_get_feature, \
+	_dom_node_set_user_data, \
+	_dom_node_get_user_data
+
+/* Following comes the protected vtable  */
+void _dom_document_destroy(dom_node_internal *node);
+dom_exception _dom_document_copy(dom_node_internal *old, 
+		dom_node_internal **copy);
+
+#define DOM_DOCUMENT_PROTECT_VTABLE \
+	_dom_document_destroy, \
+	_dom_document_copy
+
+
+/*---------------------------- Helper functions ---------------------------*/
+
+/* Try to destroy the document:
+ * When the refcnt is zero and the pending list is empty, we can destroy this
+ * document. */
+void _dom_document_try_destroy(dom_document *doc);
+
+/* Get a nodelist, creating one if necessary */
+dom_exception _dom_document_get_nodelist(dom_document *doc,
+		nodelist_type type, dom_node_internal *root,
+		dom_string *tagname, dom_string *namespace,
+		dom_string *localname, dom_nodelist **list);
+/* Remove a nodelist */
+void _dom_document_remove_nodelist(dom_document *doc, dom_nodelist *list);
+
+/* Find element with certain ID in the subtree rooted at root */
+dom_exception _dom_find_element_by_id(dom_node_internal *root, 
+		dom_string *id, dom_element **result);
+
+/* Set the ID attribute name of this document */
+void _dom_document_set_id_name(dom_document *doc, dom_string *name);
+
+#define _dom_document_get_id_name(d) (d->id_name)
+
+#endif
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/libdom/src/core/element.c browser/libdom/src/core/element.c
--- upstream/libdom/src/core/element.c	2026-09-20 23:27:12.695343544 +0100
+++ browser/libdom/src/core/element.c	2026-09-20 23:19:17.307393598 +0100
@@ -19,18 +19,19 @@
 #include <dom/core/node.h>
 #include <dom/core/string.h>
 #include <dom/core/document.h>
-#include <dom/events/events.h>
+#include <dom/events/events.h>
 
-#include "core/attr.h"
-#include "core/document.h"
-#include "core/element.h"
-#include "core/node.h"
-#include "core/namednodemap.h"
-#include "utils/validate.h"
-#include "utils/namespace.h"
-#include "utils/utils.h"
-#include "utils/list.h"
-#include "events/mutation_event.h"
+#include "core/attr.h"
+#include "core/document.h"
+#include "core/element.h"
+#include "core/node.h"
+#include "core/namednodemap.h"
+#include "utils/validate.h"
+#include "utils/namespace.h"
+#include "utils/utils.h"
+#include "utils/list.h"
+#include "events/mutation_event.h"
+#include "events/dispatch.h"
 
 const struct dom_element_vtable _dom_element_vtable = {
 	{
@@ -1676,6 +1677,52 @@
 		if (err != DOM_NO_ERR)
 			return err;
 
+		/* ==== MACSURF LOCAL PATCH (fixes1042, #310) ==================
+		 * grep MACSURF LOCAL PATCH before updating vendored libdom --
+		 * this is not upstream and a vendor bump will silently revert it.
+		 *
+		 * REFRESH THE CACHED CLASS LIST ON AN ATTRIBUTE *UPDATE*.
+		 *
+		 * dom_element caches the parsed class list in ele->classes, and
+		 * the selector engine matches against THAT, not against the
+		 * attribute string. Upstream calls _dom_element_create_classes
+		 * from exactly one place -- _dom_element_attr_list_node_create
+		 * -- which runs only when the class attribute node is FIRST
+		 * created. This branch (attribute exists, replace its value)
+		 * never refreshed it, so every class change after the first was
+		 * invisible to selector matching, while getAttribute('class')
+		 * kept returning the new value correctly.
+		 *
+		 * Hardware signature: el.className='a' restyles; el.className='b'
+		 * then does nothing. That is every classList.add/remove toggle
+		 * after the first one -- the dominant way real sites express
+		 * visual state (is-open, active, selected, has-js).
+		 *
+		 * Latent for the entire life of the port: nothing re-set class at
+		 * runtime until JS became real. Same shape as the fixes1005
+		 * double-fire -- an upstream bug that only activates once we
+		 * become its first real caller.
+		 *
+		 * Reads the value back off the attr (rather than using `value`
+		 * directly) to mirror the upstream call site exactly, including
+		 * its NUL-termination assumption. A NULL value clears the cache
+		 * via the "" path rather than dereferencing NULL, which
+		 * _dom_element_create_classes would do on its first loop.
+		 * ============================================================= */
+		if (namespace == NULL && doc != NULL &&
+				dom_string_isequal(name, doc->class_string)) {
+			dom_string *cls = NULL;
+			err = dom_attr_get_value(match->attr, &cls);
+			if (err == DOM_NO_ERR) {
+				err = _dom_element_create_classes(element,
+					(cls != NULL) ? dom_string_data(cls) : "");
+				if (cls != NULL)
+					dom_string_unref(cls);
+				if (err != DOM_NO_ERR)
+					return err;
+			}
+		}
+
 		success = true;
 		err = _dom_dispatch_subtree_modified_event(doc,
 				(dom_event_target *) e, &success);
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/libdom/src/core/element_i.h browser/libdom/src/core/element_i.h
--- upstream/libdom/src/core/element_i.h	1970-01-01 01:00:00.000000000 +0100
+++ browser/libdom/src/core/element_i.h	2026-09-20 23:19:17.307572174 +0100
@@ -0,0 +1,237 @@
+/*
+ * This file is part of libdom.
+ * Licensed under the MIT License,
+ *                http://www.opensource.org/licenses/mit-license.php
+ * Copyright 2007 John-Mark Bell <jmb@netsurf-browser.org>
+ */
+
+#ifndef dom_internal_core_element_h_
+#define dom_internal_core_element_h_
+
+/* stdbool.h removed for CW8 C89 - prefix provides bool */
+
+#include <dom/core/element.h>
+
+#include "core/node.h"
+
+struct dom_document;
+struct dom_element;
+struct dom_namednodemap;
+struct dom_node;
+struct dom_attr;
+struct dom_attr_list;
+struct dom_type_info;
+struct dom_hash_table;
+
+/**
+ * DOM element node
+ */
+struct dom_element {
+	struct dom_node_internal base;		/**< Base node */
+
+	struct dom_attr_list *attributes;	/**< Element attributes */
+
+	dom_string *id_ns;	/**< The id attribute's namespace */
+
+	dom_string *id_name; 	/**< The id attribute's name */
+
+	struct dom_type_info *schema_type_info;	/**< Type information */
+
+	lwc_string **classes;
+	uint32_t n_classes;
+};
+
+dom_exception _dom_element_create(struct dom_document *doc,
+		dom_string *name, dom_string *namespace,
+		dom_string *prefix, struct dom_element **result);
+
+dom_exception _dom_element_initialise(struct dom_document *doc,
+		struct dom_element *el, dom_string *name, 
+		dom_string *namespace, dom_string *prefix);
+
+void _dom_element_finalise(struct dom_element *ele);
+
+void _dom_element_destroy(struct dom_element *element);
+
+
+/* The virtual functions of dom_element */
+dom_exception _dom_element_get_tag_name(struct dom_element *element,
+		dom_string **name);
+dom_exception _dom_element_get_attribute(struct dom_element *element,
+		dom_string *name, dom_string **value);
+dom_exception _dom_element_set_attribute(struct dom_element *element,
+		dom_string *name, dom_string *value);
+dom_exception _dom_element_remove_attribute(struct dom_element *element, 
+		dom_string *name);
+dom_exception _dom_element_get_attribute_node(struct dom_element *element, 
+		dom_string *name, struct dom_attr **result);
+dom_exception _dom_element_set_attribute_node(struct dom_element *element, 
+		struct dom_attr *attr, struct dom_attr **result);
+dom_exception _dom_element_remove_attribute_node(struct dom_element *element, 
+		struct dom_attr *attr, struct dom_attr **result);
+dom_exception _dom_element_get_elements_by_tag_name(
+		struct dom_element *element, dom_string *name,
+		struct dom_nodelist **result);
+
+dom_exception _dom_element_get_attribute_ns(struct dom_element *element, 
+		dom_string *namespace, dom_string *localname, 
+		dom_string **value);
+dom_exception _dom_element_set_attribute_ns(struct dom_element *element,
+		dom_string *namespace, dom_string *qname,
+		dom_string *value);
+dom_exception _dom_element_remove_attribute_ns(struct dom_element *element,
+		dom_string *namespace, dom_string *localname);
+dom_exception _dom_element_get_attribute_node_ns(struct dom_element *element,
+		dom_string *namespace, dom_string *localname, 
+		struct dom_attr **result);
+dom_exception _dom_element_set_attribute_node_ns(struct dom_element *element, 
+		struct dom_attr *attr, struct dom_attr **result);
+dom_exception _dom_element_get_elements_by_tag_name_ns(
+		struct dom_element *element, dom_string *namespace, 
+		dom_string *localname, struct dom_nodelist **result);
+dom_exception _dom_element_has_attribute(struct dom_element *element,
+		dom_string *name, bool *result);
+dom_exception _dom_element_has_attribute_ns(struct dom_element *element,
+		dom_string *namespace, dom_string *localname, 
+		bool *result);
+dom_exception _dom_element_get_schema_type_info(struct dom_element *element, 
+		struct dom_type_info **result);
+dom_exception _dom_element_set_id_attribute(struct dom_element *element, 
+		dom_string *name, bool is_id);
+dom_exception _dom_element_set_id_attribute_ns(struct dom_element *element, 
+		dom_string *namespace, dom_string *localname, 
+		bool is_id);
+dom_exception _dom_element_set_id_attribute_node(struct dom_element *element,
+		struct dom_attr *id_attr, bool is_id);
+dom_exception _dom_element_get_classes(struct dom_element *element,
+		lwc_string ***classes, uint32_t *n_classes);
+dom_exception _dom_element_has_class(struct dom_element *element,
+		lwc_string *name, bool *match);
+
+#define DOM_ELEMENT_VTABLE \
+	_dom_element_get_tag_name, \
+	_dom_element_get_attribute, \
+	_dom_element_set_attribute, \
+	_dom_element_remove_attribute, \
+	_dom_element_get_attribute_node, \
+	_dom_element_set_attribute_node, \
+	_dom_element_remove_attribute_node, \
+	_dom_element_get_elements_by_tag_name, \
+	_dom_element_get_attribute_ns, \
+	_dom_element_set_attribute_ns, \
+	_dom_element_remove_attribute_ns, \
+	_dom_element_get_attribute_node_ns, \
+	_dom_element_set_attribute_node_ns, \
+	_dom_element_get_elements_by_tag_name_ns, \
+	_dom_element_has_attribute, \
+	_dom_element_has_attribute_ns, \
+	_dom_element_get_schema_type_info, \
+	_dom_element_set_id_attribute, \
+	_dom_element_set_id_attribute_ns, \
+	_dom_element_set_id_attribute_node, \
+	_dom_element_get_classes, \
+	_dom_element_has_class
+
+/* Overloading dom_node functions */
+dom_exception _dom_element_get_attributes(dom_node_internal *node,
+		struct dom_namednodemap **result);
+dom_exception _dom_element_has_attributes(dom_node_internal *node,
+		bool *result);
+dom_exception _dom_element_normalize(dom_node_internal *node);
+dom_exception _dom_element_lookup_prefix(dom_node_internal *node,
+		dom_string *namespace, dom_string **result);
+dom_exception _dom_element_is_default_namespace(dom_node_internal *node,
+		dom_string *namespace, bool *result);
+dom_exception _dom_element_lookup_namespace(dom_node_internal *node,
+		dom_string *prefix, dom_string **result);
+#define DOM_NODE_VTABLE_ELEMENT \
+	_dom_node_try_destroy, \
+	_dom_node_get_node_name, \
+	_dom_node_get_node_value, \
+	_dom_node_set_node_value, \
+	_dom_node_get_node_type, \
+	_dom_node_get_parent_node, \
+	_dom_node_get_child_nodes, \
+	_dom_node_get_first_child, \
+	_dom_node_get_last_child, \
+	_dom_node_get_previous_sibling, \
+	_dom_node_get_next_sibling, \
+	_dom_element_get_attributes, /*overload*/\
+	_dom_node_get_owner_document, \
+	_dom_node_insert_before, \
+	_dom_node_replace_child, \
+	_dom_node_remove_child, \
+	_dom_node_append_child, \
+	_dom_node_has_child_nodes, \
+	_dom_node_clone_node, \
+	_dom_node_normalize, \
+	_dom_node_is_supported, \
+	_dom_node_get_namespace, \
+	_dom_node_get_prefix, \
+	_dom_node_set_prefix, \
+	_dom_node_get_local_name, \
+	_dom_element_has_attributes, /*overload*/\
+	_dom_node_get_base, \
+	_dom_node_compare_document_position, \
+	_dom_node_get_text_content, \
+	_dom_node_set_text_content, \
+	_dom_node_is_same, \
+	_dom_element_lookup_prefix, /*overload*/\
+	_dom_element_is_default_namespace, /*overload*/\
+	_dom_element_lookup_namespace, /*overload*/\
+	_dom_node_is_equal, \
+	_dom_node_get_feature, \
+	_dom_node_set_user_data, \
+	_dom_node_get_user_data
+
+/**
+ * The internal used vtable for element
+ */
+struct dom_element_protected_vtable {
+	struct dom_node_protect_vtable base;
+
+	dom_exception (*dom_element_parse_attribute)(dom_element *ele,
+			dom_string *name, dom_string *value,
+			dom_string **parsed);
+			/**< Called by dom_attr_set_value, and used to check
+			 *   whether the new attribute value is valid and 
+			 *   return a valid on if it is not
+			 */
+};
+
+typedef struct dom_element_protected_vtable dom_element_protected_vtable;
+
+/* MacSurf: body in libdom_c89_helpers.c */
+extern dom_exception dom_element_parse_attribute_impl(dom_element *ele,
+		dom_string *name, dom_string *value, dom_string **parsed);
+#define dom_element_parse_attribute(e, n, v, p) dom_element_parse_attribute_impl( \
+		(dom_element *) (e), (dom_string *) (n), \
+		(dom_string *) (v), (dom_string **) (p))
+
+
+/* The protected virtual function */
+dom_exception _dom_element_parse_attribute(dom_element *ele,
+		dom_string *name, dom_string *value,
+		dom_string **parsed);
+void __dom_element_destroy(dom_node_internal *node);
+dom_exception _dom_element_copy(dom_node_internal *old, 
+		dom_node_internal **copy);
+
+#define DOM_ELEMENT_PROTECT_VTABLE \
+	_dom_element_parse_attribute
+
+#define DOM_NODE_PROTECT_VTABLE_ELEMENT \
+	__dom_element_destroy, \
+	_dom_element_copy
+
+/* Helper functions*/
+dom_exception _dom_element_copy_internal(dom_element *old,
+		dom_element *new);
+#define dom_element_copy_internal(o, n) _dom_element_copy_internal( \
+		(dom_element *) (o), (dom_element *) (n))
+
+dom_exception _dom_element_get_id(struct dom_element *ele, dom_string **id);
+
+extern const struct dom_element_vtable _dom_element_vtable;
+
+#endif
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/libdom/src/core/namednodemap_i.h browser/libdom/src/core/namednodemap_i.h
--- upstream/libdom/src/core/namednodemap_i.h	1970-01-01 01:00:00.000000000 +0100
+++ browser/libdom/src/core/namednodemap_i.h	2026-09-20 23:19:17.308382521 +0100
@@ -0,0 +1,70 @@
+/*
+ * This file is part of libdom.
+ * Licensed under the MIT License,
+ *                http://www.opensource.org/licenses/mit-license.php
+ * Copyright 2007 John-Mark Bell <jmb@netsurf-browser.org>
+ */
+
+#ifndef dom_internal_core_namednodemap_h_
+#define dom_internal_core_namednodemap_h_
+
+/* stdbool removed for CW8 -- prefix provides bool */
+
+#include <dom/core/namednodemap.h>
+#include <dom/core/node.h>
+
+struct dom_document;
+struct dom_node;
+struct dom_namednodemap;
+
+struct nnm_operation {
+	dom_exception (*namednodemap_get_length)(void *priv,
+			dom_ulong *length);
+
+	dom_exception (*namednodemap_get_named_item)(void *priv,
+			dom_string *name, struct dom_node **node);
+
+	dom_exception (*namednodemap_set_named_item)(void *priv,
+			struct dom_node *arg, struct dom_node **node);
+
+	dom_exception (*namednodemap_remove_named_item)(
+			void *priv, dom_string *name,
+			struct dom_node **node);
+
+	dom_exception (*namednodemap_item)(void *priv,
+			dom_ulong index, struct dom_node **node);
+
+	dom_exception (*namednodemap_get_named_item_ns)(
+			void *priv, dom_string *namespace,
+			dom_string *localname, struct dom_node **node);
+
+	dom_exception (*namednodemap_set_named_item_ns)(
+			void *priv, struct dom_node *arg,
+			struct dom_node **node);
+
+	dom_exception (*namednodemap_remove_named_item_ns)(
+			void *priv, dom_string *namespace,
+			dom_string *localname, struct dom_node **node);
+
+	void (*namednodemap_destroy)(void *priv);
+
+	bool (*namednodemap_equal)(void *p1, void *p2);
+};
+
+/* Create a namednodemap */
+dom_exception _dom_namednodemap_create(struct dom_document *doc,
+		void *priv, struct nnm_operation *opt,
+		struct dom_namednodemap **map);
+
+/* Update the private data */
+void _dom_namednodemap_update(struct dom_namednodemap *map, void *priv);
+
+/* Test whether two maps are equal */
+bool _dom_namednodemap_equal(struct dom_namednodemap *m1, 
+		struct dom_namednodemap *m2);
+
+#define dom_namednodemap_equal(m1, m2) _dom_namednodemap_equal( \
+		(struct dom_namednodemap *) (m1), \
+		(struct dom_namednodemap *) (m2))
+
+#endif
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/libdom/src/core/node_i.h browser/libdom/src/core/node_i.h
--- upstream/libdom/src/core/node_i.h	1970-01-01 01:00:00.000000000 +0100
+++ browser/libdom/src/core/node_i.h	2026-09-20 23:19:17.308840445 +0100
@@ -0,0 +1,299 @@
+/*
+ * This file is part of libdom.
+ * Licensed under the MIT License,
+ *                http://www.opensource.org/licenses/mit-license.php
+ * Copyright 2007 John-Mark Bell <jmb@netsurf-browser.org>
+ */
+
+#ifndef dom_internal_core_node_h_
+#define dom_internal_core_node_h_
+
+#include <libwapcaplet/libwapcaplet.h>
+
+#include <dom/core/node.h>
+#include <dom/functypes.h>
+
+#include "events/event_target.h"
+#include "events/mutation_event.h"
+
+#include "utils/list.h"
+
+/**
+ * User data context attached to a DOM node
+ */
+struct dom_user_data {
+	dom_string *key;		/**< Key for data */
+	void *data;			/**< Client-specific data */
+	dom_user_data_handler handler;	/**< Callback function */
+
+	struct dom_user_data *next;	/**< Next in list */
+	struct dom_user_data *prev;	/**< Previous in list */
+};
+typedef struct dom_user_data dom_user_data;
+
+/**
+ * The internally used virtual function table.
+ */
+typedef struct dom_node_protect_vtable {
+
+	void (*destroy)(dom_node_internal *n);
+					/**< The destroy virtual function, it 
+					 * should be private to client */
+	dom_exception (*copy)(dom_node_internal *old, dom_node_internal **copy);
+				/**< Copy the old to new as well as 
+				 * all its attributes, but not its children */
+} dom_node_protect_vtable; 
+
+/**
+ * The real DOM node object
+ *
+ * DOM nodes are reference counted
+ */
+struct dom_node_internal {
+	struct dom_node base;		/**< The vtable base */
+	const void *vtable;		/**< The protected vtable */
+
+	dom_string *name;		/**< Node name (this is the local part
+		 			 * of a QName in the cases where a
+		 			 * namespace exists) */
+	dom_string *value;		/**< Node value */
+	dom_node_type type;		/**< Node type */
+	dom_node_internal *parent;	/**< Parent node */
+	dom_node_internal *first_child;	/**< First child node */
+	dom_node_internal *last_child;	/**< Last child node */
+	dom_node_internal *previous;	/**< Previous sibling */
+	dom_node_internal *next;		/**< Next sibling */
+
+	struct dom_document *owner;	/**< Owning document */
+
+	dom_string *namespace;		/**< Namespace URI */
+	dom_string *prefix;		/**< Namespace prefix */
+
+	struct dom_user_data *user_data;	/**< User data list */
+
+	struct list_entry pending_list; /**< The document delete pending list */
+
+	dom_event_target_internal eti;	/**< The EventTarget interface */
+};
+
+dom_node_internal * _dom_node_create(void);
+
+dom_exception _dom_node_initialise(struct dom_node_internal *node,
+		struct dom_document *doc, dom_node_type type,
+		dom_string *name, dom_string *value,
+		dom_string *namespace, dom_string *prefix);
+
+void _dom_node_finalise(dom_node_internal *node);
+
+bool _dom_node_readonly_owner(const dom_node_internal *node);
+bool _dom_node_readonly(const dom_node_internal *node);
+
+/* Event Target implementation */
+dom_exception _dom_node_add_event_listener(dom_event_target *et,
+		dom_string *type, struct dom_event_listener *listener, 
+		bool capture);
+dom_exception _dom_node_remove_event_listener(dom_event_target *et,
+		dom_string *type, struct dom_event_listener *listener, 
+		bool capture);
+dom_exception _dom_node_add_event_listener_ns(dom_event_target *et,
+		dom_string *namespace, dom_string *type, 
+		struct dom_event_listener *listener, bool capture);
+dom_exception _dom_node_remove_event_listener_ns(dom_event_target *et,
+		dom_string *namespace, dom_string *type, 
+		struct dom_event_listener *listener, bool capture);
+dom_exception _dom_node_dispatch_event(dom_event_target *et,
+		struct dom_event *evt, bool *success);
+
+/* The DOM Node's vtable methods */
+dom_exception _dom_node_get_node_name(dom_node_internal *node,
+		dom_string **result);
+dom_exception _dom_node_get_node_value(dom_node_internal *node,
+		dom_string **result);
+dom_exception _dom_node_set_node_value(dom_node_internal *node,
+		dom_string *value);
+dom_exception _dom_node_get_node_type(dom_node_internal *node,
+		dom_node_type *result);
+dom_exception _dom_node_get_parent_node(dom_node_internal *node,
+		dom_node_internal **result);
+dom_exception _dom_node_get_child_nodes(dom_node_internal *node,
+		struct dom_nodelist **result);
+dom_exception _dom_node_get_first_child(dom_node_internal *node,
+		dom_node_internal **result);
+dom_exception _dom_node_get_last_child(dom_node_internal *node,
+		dom_node_internal **result);
+dom_exception _dom_node_get_previous_sibling(dom_node_internal *node,
+		dom_node_internal **result);
+dom_exception _dom_node_get_next_sibling(dom_node_internal *node,
+		dom_node_internal **result);
+dom_exception _dom_node_get_attributes(dom_node_internal *node,
+		struct dom_namednodemap **result);
+dom_exception _dom_node_get_owner_document(dom_node_internal *node,
+		struct dom_document **result);
+dom_exception _dom_node_insert_before(dom_node_internal *node,
+		dom_node_internal *new_child, dom_node_internal *ref_child,
+		dom_node_internal **result);
+dom_exception _dom_node_replace_child(dom_node_internal *node,
+		dom_node_internal *new_child, dom_node_internal *old_child,
+		dom_node_internal **result);
+dom_exception _dom_node_remove_child(dom_node_internal *node,
+		dom_node_internal *old_child,
+		dom_node_internal **result);
+dom_exception _dom_node_append_child(dom_node_internal *node,
+		dom_node_internal *new_child,
+		dom_node_internal **result);
+dom_exception _dom_node_has_child_nodes(dom_node_internal *node, bool *result);
+dom_exception _dom_node_clone_node(dom_node_internal *node, bool deep,
+		dom_node_internal **result);
+dom_exception _dom_node_normalize(dom_node_internal *node);
+dom_exception _dom_node_is_supported(dom_node_internal *node,
+		dom_string *feature, dom_string *version,
+		bool *result);
+dom_exception _dom_node_get_namespace(dom_node_internal *node,
+		dom_string **result);
+dom_exception _dom_node_get_prefix(dom_node_internal *node,
+		dom_string **result);
+dom_exception _dom_node_set_prefix(dom_node_internal *node,
+		dom_string *prefix);
+dom_exception _dom_node_get_local_name(dom_node_internal *node,
+		dom_string **result);
+dom_exception _dom_node_has_attributes(dom_node_internal *node, bool *result);
+dom_exception _dom_node_get_base(dom_node_internal *node,
+		dom_string **result);
+dom_exception _dom_node_compare_document_position(dom_node_internal *node,
+		dom_node_internal *other, uint16_t *result);
+dom_exception _dom_node_get_text_content(dom_node_internal *node,
+		dom_string **result);
+dom_exception _dom_node_set_text_content(dom_node_internal *node,
+		dom_string *content);
+dom_exception _dom_node_is_same(dom_node_internal *node,
+		dom_node_internal *other, bool *result);
+dom_exception _dom_node_lookup_prefix(dom_node_internal *node,
+		dom_string *namespace, dom_string **result);
+dom_exception _dom_node_is_default_namespace(dom_node_internal *node,
+		dom_string *namespace, bool *result);
+dom_exception _dom_node_lookup_namespace(dom_node_internal *node,
+		dom_string *prefix, dom_string **result);
+dom_exception _dom_node_is_equal(dom_node_internal *node,
+		dom_node_internal *other, bool *result);
+dom_exception _dom_node_get_feature(dom_node_internal *node,
+		dom_string *feature, dom_string *version,
+		void **result);
+dom_exception _dom_node_set_user_data(dom_node_internal *node,
+		dom_string *key, void *data,
+		dom_user_data_handler handler, void **result);
+dom_exception _dom_node_get_user_data(const dom_node_internal *node,
+		const dom_string *key, void **result);
+
+#define DOM_NODE_EVENT_TARGET_VTABLE \
+	_dom_node_add_event_listener, \
+	_dom_node_remove_event_listener, \
+	_dom_node_dispatch_event, \
+	_dom_node_add_event_listener_ns, \
+	_dom_node_remove_event_listener_ns
+
+#define DOM_NODE_VTABLE \
+	_dom_node_try_destroy, \
+	_dom_node_get_node_name,	  \
+	_dom_node_get_node_value, \
+	_dom_node_set_node_value, \
+	_dom_node_get_node_type, \
+	_dom_node_get_parent_node, \
+	_dom_node_get_child_nodes, \
+	_dom_node_get_first_child, \
+	_dom_node_get_last_child, \
+	_dom_node_get_previous_sibling, \
+	_dom_node_get_next_sibling, \
+	_dom_node_get_attributes, \
+	_dom_node_get_owner_document, \
+	_dom_node_insert_before, \
+	_dom_node_replace_child, \
+	_dom_node_remove_child, \
+	_dom_node_append_child, \
+	_dom_node_has_child_nodes, \
+	_dom_node_clone_node, \
+	_dom_node_normalize, \
+	_dom_node_is_supported, \
+	_dom_node_get_namespace, \
+	_dom_node_get_prefix, \
+	_dom_node_set_prefix, \
+	_dom_node_get_local_name, \
+	_dom_node_has_attributes, \
+	_dom_node_get_base, \
+	_dom_node_compare_document_position, \
+	_dom_node_get_text_content, \
+	_dom_node_set_text_content, \
+	_dom_node_is_same, \
+	_dom_node_lookup_prefix, \
+	_dom_node_is_default_namespace, \
+	_dom_node_lookup_namespace, \
+	_dom_node_is_equal, \
+	_dom_node_get_feature, \
+	_dom_node_set_user_data, \
+	_dom_node_get_user_data
+
+
+/* Following comes the protected vtable */
+void _dom_node_destroy(struct dom_node_internal *node);
+dom_exception _dom_node_copy(struct dom_node_internal *old, 
+		struct dom_node_internal **copy);
+
+#define DOM_NODE_PROTECT_VTABLE \
+	_dom_node_destroy, \
+	_dom_node_copy
+
+
+/* MacSurf: bodies in libdom_c89_helpers.c */
+extern void dom_node_destroy_impl(struct dom_node_internal *node);
+extern dom_exception dom_node_copy_impl(struct dom_node_internal *old,
+		struct dom_node_internal **copy);
+#define dom_node_destroy(n) dom_node_destroy_impl((dom_node_internal *) (n))
+#define dom_node_copy(o,c) dom_node_copy_impl((dom_node_internal *) (o), \
+		(dom_node_internal **) (c))
+
+/* Following are some helper functions */
+dom_exception _dom_node_copy_internal(dom_node_internal *old, 
+		dom_node_internal *new);
+#define dom_node_copy_internal(o, n) _dom_node_copy_internal( \
+		(dom_node_internal *) (o), (dom_node_internal *) (n))
+
+#define dom_node_get_owner(n) ((dom_node_internal *) (n))->owner
+
+#define dom_node_set_owner(n, d) ((dom_node_internal *) (n))->owner = \
+		(struct dom_document *) (d)
+
+#define dom_node_get_parent(n) ((dom_node_internal *) (n))->parent
+
+#define dom_node_set_parent(n, p) ((dom_node_internal *) (n))->parent = \
+		(dom_node_internal *) (p)
+
+#define dom_node_get_refcount(n) ((dom_node_internal *) (n))->refcnt
+
+dom_exception _dom_merge_adjacent_text(dom_node_internal *p,
+		dom_node_internal *n);
+
+/* Try to destroy the node, if its refcnt is not zero, then append it to the
+ * owner document's pending list */
+dom_exception _dom_node_try_destroy(dom_node_internal *node);
+
+/* To add some node to the pending list */
+void _dom_node_mark_pending(dom_node_internal *node);
+#define dom_node_mark_pending(n) _dom_node_mark_pending(\
+		(dom_node_internal *) (n))
+/* To remove the node from the pending list, this may happen when
+ * a node is removed and then appended to another parent */
+void _dom_node_remove_pending(dom_node_internal *node);
+#define dom_node_remove_pending(n) _dom_node_remove_pending(\
+		(dom_node_internal *) (n))
+
+dom_exception _dom_node_dispatch_node_change_event(dom_document *doc,
+		dom_node_internal *node, dom_node_internal *related,
+		dom_mutation_type change, bool *success);
+#define dom_node_dispatch_node_change_event( \
+		doc, node, related, change, success) \
+	_dom_node_dispatch_node_change_event((dom_document *) (doc), \
+			(dom_node_internal *) (node), \
+			(dom_node_internal *) (related), \
+			(dom_mutation_type) (change), \
+			(bool *) (success))
+
+#endif
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/libdom/src/core/nodelist_i.h browser/libdom/src/core/nodelist_i.h
--- upstream/libdom/src/core/nodelist_i.h	1970-01-01 01:00:00.000000000 +0100
+++ browser/libdom/src/core/nodelist_i.h	2026-09-20 23:19:17.309141820 +0100
@@ -0,0 +1,45 @@
+/*
+ * This file is part of libdom.
+ * Licensed under the MIT License,
+ *                http://www.opensource.org/licenses/mit-license.php
+ * Copyright 2007 John-Mark Bell <jmb@netsurf-browser.org>
+ */
+
+#ifndef dom_internal_core_nodelist_h_
+#define dom_internal_core_nodelist_h_
+
+/* stdbool.h removed for CW8 C89 -- prefix provides bool */
+
+#include <dom/core/nodelist.h>
+
+struct dom_document;
+struct dom_node;
+struct dom_nodelist;
+
+/**
+ * The NodeList type
+ */
+typedef enum { 
+	DOM_NODELIST_CHILDREN,
+	DOM_NODELIST_BY_NAME,
+	DOM_NODELIST_BY_NAMESPACE,
+	DOM_NODELIST_BY_NAME_CASELESS,
+	DOM_NODELIST_BY_NAMESPACE_CASELESS
+} nodelist_type;
+
+/* Create a nodelist */
+dom_exception _dom_nodelist_create(struct dom_document *doc, nodelist_type type,
+		struct dom_node_internal *root, dom_string *tagname,
+		dom_string *namespace, dom_string *localname,
+		struct dom_nodelist **list);
+
+/* Match a nodelist instance against a set of nodelist creation parameters */
+bool _dom_nodelist_match(struct dom_nodelist *list, nodelist_type type,
+		struct dom_node_internal *root, dom_string *tagname, 
+		dom_string *namespace, dom_string *localname);
+
+bool _dom_nodelist_equal(struct dom_nodelist *l1, struct dom_nodelist *l2);
+#define dom_nodelist_equal(l1, l2) _dom_nodelist_equal( \
+		(struct dom_nodelist *) (l1), (struct dom_nodelist *) (l2))
+
+#endif
diff -urN upstream/libdom/src/core/string.c browser/libdom/src/core/string.c
--- upstream/libdom/src/core/string.c
+++ browser/libdom/src/core/string.c
@@ -14,9 +14,11 @@
 
 #include <parserutils/charset/utf8.h>
 
-#include "core/string.h"
-#include "core/document.h"
-#include "utils/utils.h"
+#include "core/string.h"
+#include "core/document.h"
+#include "utils/utils.h"
+
+extern int macsurf_ptr_is_heap(const void *);
 
 /**
  * Type of a DOM string
@@ -178,8 +180,24 @@
 	if (istr->type != DOM_STRING_INTERNED) {
 		lwc_string *ret;
 		lwc_error lerr;
-
-		lerr = lwc_intern_string((const char *) istr->data.cdata.ptr, 
+		/* fixes446f: guard against insane length only. Pointer check
+		 * (pa_ >= 0x28000000) removed: PEF string literals live at
+		 * 0x3E... (code space on OS 9) and are valid CDATA sources.
+		 * The corrupt-sched_entry case has sched_entry.p (~0x07000000
+		 * = 117 MB) aliased as cdata.len, which the length check
+		 * catches cleanly. */
+		if (istr->data.cdata.len > 1048576UL) {
+			extern void macsurf_debug_log_writef(
+				const char *fmt, ...);
+			macsurf_debug_log_writef(
+				"fixes446f INTERN: bad len=%ld ptr=%p "
+				"str=%p",
+				(long)istr->data.cdata.len,
+				(void *)istr->data.cdata.ptr,
+				(void *)istr);
+			return DOM_NO_MEM_ERR;
+		}
+		lerr = lwc_intern_string((const char *) istr->data.cdata.ptr,
 				istr->data.cdata.len, &ret);
 		if (lerr != lwc_error_ok) {
 			return _dom_exception_from_lwc_error(lerr);
@@ -218,7 +236,7 @@
 
 	if (is1->type == DOM_STRING_INTERNED && 
 			is2->type == DOM_STRING_INTERNED) {
-		bool match;
+		unsigned char match;   /* shim isequal takes unsigned char * */
 
 		(void) lwc_string_isequal(is1->data.intern, is2->data.intern,
 			&match);
@@ -267,7 +285,7 @@
 
 	if (is1->type == DOM_STRING_INTERNED && 
 			is2->type == DOM_STRING_INTERNED) {
-		bool match;
+		unsigned char match;   /* shim isequal takes unsigned char * */
 
 		if (lwc_string_caseless_isequal(is1->data.intern,
 				is2->data.intern, &match) != lwc_error_ok)
@@ -315,7 +333,7 @@
 		return false;
 
 	if (is1->type == DOM_STRING_INTERNED) {
-		bool match;
+		unsigned char match;   /* shim isequal takes unsigned char * */
 
 		(void) lwc_string_isequal(is1->data.intern, s2, &match);
 
@@ -352,7 +370,7 @@
 		return false;
 
 	if (is1->type == DOM_STRING_INTERNED) {
-		bool match;
+		unsigned char match;   /* shim isequal takes unsigned char * */
 
 		if (lwc_string_caseless_isequal(is1->data.intern, s2, &match) != lwc_error_ok)
 			return false;
@@ -896,23 +914,116 @@
 const char *dom_string_data(const dom_string *str)
 {
 	dom_string_internal *istr = (void *) str;
+	/* fixes490 - NULL guard. A NULL dom_string here dereferences
+	 * address 0: istr->type reads low memory, and if it happens to be
+	 * DOM_STRING_CDATA (== 0) the code reads data.cdata.ptr at struct
+	 * offset 4 - a low-memory global that, on this G3, points into the
+	 * intern table and renders an unrelated interned string (the
+	 * "data-xf-init" attribute name) as page text. The fixes446g log
+	 * showed exactly this: ptr=3E0006B8 str=00000000. Return "" so a
+	 * NULL string yields empty text instead of garbage. */
+	if (istr == NULL)
+		return "";
 	if (istr->type == DOM_STRING_CDATA) {
+		/* fixes446g: guard against corrupt CDATA ptr from sched_entry
+		 * UAF (sched_entry.callback, a code-space fn-ptr at 0x3E...,
+		 * aliases data.cdata.ptr at struct offset 4).  A legitimate
+		 * CDATA ptr is always malloc'd heap (< 0x28000000).  Return ""
+		 * so callers get an empty string rather than a strlen crash. */
+		{
+			unsigned long pa_ =
+				(unsigned long)istr->data.cdata.ptr;
+			/* fixes489 (Option A) - widen the recycled-struct
+			 * guard. The original check only rejected a
+			 * code-space pointer. A freed dom_string struct that
+			 * the scheduler reused can also leave data.cdata.ptr
+			 * pointing at heap (passing the range test) while
+			 * data.cdata.len holds garbage. A legitimate CDATA
+			 * length is bounded (text nodes here never exceed the
+			 * 1 MB intern cap used a few lines down); a recycled
+			 * struct's len is typically a huge or wrapped value.
+			 * Reject both an out-of-range ptr and an absurd len so
+			 * the caller gets "" instead of reading recycled
+			 * memory as text. */
+			if (!macsurf_ptr_is_heap((const void *)istr->data.cdata.ptr) ||
+			    istr->data.cdata.len > 1048576UL) {
+				extern void macsurf_debug_log_writef(
+					const char *fmt, ...);
+				macsurf_debug_log_writef(
+					"fixes446g DATA: bad cdata ptr=%p "
+					"len=%ld str=%p",
+					(void *)pa_,
+					(long)istr->data.cdata.len,
+					(void *)str);
+				return "";
+			}
+		}
 		return (const char *) istr->data.cdata.ptr;
 	} else {
+		/* fixes446b: same guard as dom_string_tolower - a corrupt
+		 * intern pointer (UAF or write-past) causes lwc_string_data
+		 * to return garbage_ptr+24, which then crashes in memcpy /
+		 * duk_push_lstring / strlen.  Return "" on bad pointer. */
+		{
+			unsigned long ia_ = (unsigned long)istr->data.intern;
+			if (!macsurf_ptr_is_heap((const void *)istr->data.intern)) {
+				extern void macsurf_debug_log_writef(
+					const char *fmt, ...);
+				macsurf_debug_log_writef(
+					"fixes446b DATA: corrupt intern=%p "
+					"str=%p type=%d",
+					(void *)ia_,
+					(void *)str,
+					(int)istr->type);
+				return "";
+			}
+		}
 		return lwc_string_data(istr->data.intern);
 	}
 }
 
-/** Get the byte length of this dom_string 
+/** Get the byte length of this dom_string
  *
  * \param str	The dom_string object
  */
 size_t dom_string_byte_length(const dom_string *str)
 {
 	dom_string_internal *istr = (void *) str;
+	/* fixes490 - NULL guard, paired with dom_string_data. A NULL string
+	 * here would read low memory as the length and render garbage. */
+	if (istr == NULL)
+		return 0;
 	if (istr->type == DOM_STRING_CDATA) {
+		/* fixes569: same recycled-struct guard as dom_string_data's
+		 * CDATA branch (fixes446g/489). A dom_string_internal freed and
+		 * reused by the scheduler leaves data.cdata.ptr in code space
+		 * (>= 0x28000000) or data.cdata.len garbage. Consumers that size
+		 * a memcpy/scan by this length (box_construct_text box_construct.c
+		 * :1819, forms.c, css_fetcher.c) then read wild memory and crash
+		 * in a strlen-style loop (68kmla box_image path, r4 wild).
+		 * dom_string_data already returns "" for these; return 0 here so
+		 * the paired (data, len) agree and no overrun happens. */
+		{
+			unsigned long pa_ = (unsigned long)istr->data.cdata.ptr;
+			if (!macsurf_ptr_is_heap((const void *)istr->data.cdata.ptr) ||
+			    istr->data.cdata.len > 1048576UL) {
+				extern void macsurf_debug_log_writef(
+					const char *fmt, ...);
+				macsurf_debug_log_writef(
+					"fixes569 LEN: bad cdata ptr=%p len=%ld",
+					(void *)pa_,
+					(long)istr->data.cdata.len);
+				return 0;
+			}
+		}
 		return istr->data.cdata.len;
 	} else {
+		/* fixes446b: paired guard - same corrupt intern crashes here
+		 * via lwc_string_length reading str->len at bad_ptr+8. */
+		{
+			if (!macsurf_ptr_is_heap((const void *)istr->data.intern))
+				return 0;
+		}
 		return lwc_string_length(istr->data.intern);
 	}
 }
@@ -1008,7 +1119,28 @@
 		bool equal;
 		lwc_error err;
 		lwc_string *l;
-
+		/* fixes446: guard against heap-corrupt lwc_string pointer.
+		 * Crash signature: lbzu r0,0x0001(r4) with r4 in kernel
+		 * address space (>0x28000000 on 640MB Mac OS 9), because
+		 * CSTR_OF(bad_intern) = bad_intern+24 is unmapped.
+		 * Root cause is a UAF that writes garbage to
+		 * dom_string_internal.data.intern; this guard returns before
+		 * lwc__intern_caseless_string dereferences the bad pointer. */
+		{
+			unsigned long intern_addr_ =
+				(unsigned long)isource->data.intern;
+			if (!macsurf_ptr_is_heap((const void *)isource->data.intern)) {
+				extern void macsurf_debug_log_writef(
+					const char *fmt, ...);
+				macsurf_debug_log_writef(
+					"fixes446 GUARD: corrupt intern=%p "
+					"in dom_string_tolower source=%p type=%d",
+					(void *)intern_addr_,
+					(void *)source,
+					(int)isource->type);
+				return DOM_NO_MEM_ERR;
+			}
+		}
 		err = lwc_string_tolower(isource->data.intern, &l);
 		if (err != lwc_error_ok) {
 			return DOM_NO_MEM_ERR;
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/libdom/src/core/text.c browser/libdom/src/core/text.c
--- upstream/libdom/src/core/text.c	2026-09-20 23:27:12.695343544 +0100
+++ browser/libdom/src/core/text.c	2026-09-20 23:19:17.309590290 +0100
@@ -14,10 +14,10 @@
 
 #include <libwapcaplet/libwapcaplet.h>
 
-#include "core/characterdata.h"
-#include "core/document.h"
-#include "core/text.h"
-#include "utils/utils.h"
+#include "core/characterdata.h"
+#include "core/document.h"
+#include "core/text.h"
+#include "utils/utils.h"
 
 /* The virtual table for dom_text */
 const struct dom_text_vtable text_vtable = {
@@ -43,8 +43,8 @@
 	DELETE
 } walk_operation;
 typedef enum walk_order {
-	LEFT,
-	RIGHT
+	WALK_LEFT,
+	WALK_RIGHT
 } walk_order;
 
 /* Walk the logic-adjacent text in document order */
@@ -386,7 +386,7 @@
 					return err;
 
 				tmp = *ret;
-				if (order == LEFT) {
+				if (order == WALK_LEFT) {
 					if (tmp != NULL) {
 						err = dom_string_concat(data, tmp, ret);
 						if (err != DOM_NO_ERR)
@@ -395,7 +395,7 @@
 						dom_string_ref(data);
 						*ret = data;
 					}
-				} else if (order == RIGHT) {
+				} else if (order == WALK_RIGHT) {
 					if (tmp != NULL) {
 						err = dom_string_concat(tmp, data, ret);
 						if (err != DOM_NO_ERR)
@@ -428,7 +428,7 @@
 		}
 
 		p = dom_node_get_parent(node);
-		if (order == LEFT) {
+		if (order == WALK_LEFT) {
 			if (node->last_child != NULL) {
 				node = node->last_child;
 			} else if (node->previous != NULL) {
@@ -485,7 +485,7 @@
 	*ret = NULL;
 
 	/* Firstly, we look our left */
-	err = walk_logic_adjacent_text_in_order(left, opt, LEFT, ret, &cont);
+	err = walk_logic_adjacent_text_in_order(left, opt, WALK_LEFT, ret, &cont);
 	if (err != DOM_NO_ERR) {
 		if (opt == COLLECT) {
 			dom_string_unref(*ret);
@@ -525,7 +525,7 @@
 	}
 
 	/* Now, look right */
-	err = walk_logic_adjacent_text_in_order(right, opt, RIGHT, ret, &cont);
+	err = walk_logic_adjacent_text_in_order(right, opt, WALK_RIGHT, ret, &cont);
 	if (err != DOM_NO_ERR) {
 		if (opt == COLLECT) {
 			dom_string_unref(*ret);
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/libdom/src/core/text_i.h browser/libdom/src/core/text_i.h
--- upstream/libdom/src/core/text_i.h	1970-01-01 01:00:00.000000000 +0100
+++ browser/libdom/src/core/text_i.h	2026-09-20 23:19:17.309732943 +0100
@@ -0,0 +1,75 @@
+/*
+ * This file is part of libdom.
+ * Licensed under the MIT License,
+ *                http://www.opensource.org/licenses/mit-license.php
+ * Copyright 2007 John-Mark Bell <jmb@netsurf-browser.org>
+ */
+
+#ifndef dom_internal_core_text_h_
+#define dom_internal_core_text_h_
+
+/* stdbool removed for CW8 -- prefix provides bool */
+
+#include <dom/core/exceptions.h>
+#include <dom/core/text.h>
+
+#include "core/characterdata.h"
+
+struct dom_document;
+
+/**
+ * A DOM text node
+ */
+struct dom_text {
+	dom_characterdata base;	/**< Base node */
+
+	bool element_content_whitespace;	/**< This node is element
+						 * content whitespace */
+};
+
+/* Constructor and Destructor */
+dom_exception _dom_text_create(struct dom_document *doc,
+		dom_string *name, dom_string *value,
+		dom_text **result);
+
+void _dom_text_destroy(dom_text *text);
+
+dom_exception _dom_text_initialise(dom_text *text,
+		struct dom_document *doc, dom_node_type type,
+		dom_string *name, dom_string *value);
+
+void _dom_text_finalise(dom_text *text);
+
+
+/* Virtual functions for dom_text */
+dom_exception _dom_text_split_text(dom_text *text,
+		uint32_t offset, dom_text **result);
+dom_exception _dom_text_get_is_element_content_whitespace(
+		dom_text *text, bool *result);
+dom_exception _dom_text_get_whole_text(dom_text *text,
+		dom_string **result);
+dom_exception _dom_text_replace_whole_text(dom_text *text,
+		dom_string *content, dom_text **result);
+
+#define DOM_TEXT_VTABLE \
+	_dom_text_split_text, \
+	_dom_text_get_is_element_content_whitespace, \
+	_dom_text_get_whole_text, \
+	_dom_text_replace_whole_text
+
+
+/* Following comes the protected vtable  */
+void __dom_text_destroy(struct dom_node_internal *node);
+dom_exception _dom_text_copy(dom_node_internal *old, dom_node_internal **copy);
+
+#define DOM_TEXT_PROTECT_VTABLE \
+	__dom_text_destroy, \
+	_dom_text_copy
+
+extern const struct dom_text_vtable text_vtable;
+
+dom_exception _dom_text_copy_internal(dom_text *old, dom_text *new);
+#define dom_text_copy_internal(o, n) \
+		_dom_text_copy_internal((dom_text *) (o), (dom_text *) (n))
+
+#endif
diff -urN -x .git -x 'build-*' -x test -x frontends -x '*.o' -x '*.a' -x '*.d' -x '*.pc' -x build upstream/libdom/src/core/tokenlist_i.h browser/libdom/src/core/tokenlist_i.h
--- upstream/libdom/src/core/tokenlist_i.h	1970-01-01 01:00:00.000000000 +0100
+++ browser/libdom/src/core/tokenlist_i.h	2026-09-20 23:19:17.310076291 +0100
@@ -0,0 +1,18 @@
+/*
+ * This file is part of libdom.
+ * Licensed under the MIT License,
+ *                http://www.opensource.org/licenses/mit-license.php
+ * Copyright 2007 John-Mark Bell <jmb@netsurf-browser.org>
+ */
+
+#ifndef dom_internal_core_tokenlist_h_
+#define dom_internal_core_tokenlist_h_
+
+/* stdbool removed for CW8 -- prefix provides bool */
+
+#include <dom/core/tokenlist.h>
+
+struct dom_element;
+struct dom_string;
+
+#endif
